src/poly/cse.rs

```diff
@@ -1,134 +1,92 @@
-use rand_chacha::{rand_core::SeedableRng, ChaCha20Rng};
-
 use crate::field::Field;
 
-use std::{
-    collections::HashMap,
-    fmt::Debug,
-    hash::Hash,
-    rc::{Rc, Weak},
-};
-
-use super::{Expr, HashResult, VarAssignments};
-
-/// Common Subexpression Elimination - takes a collection of expr
-pub fn cse<F: Field + Hash, V: Debug + Clone + Eq + Hash>(
-    exprs: Vec<Expr<F, V, ()>>,
-    queriables: &Vec<V>,
-) -> Vec<Rc<Expr<F, V, HashResult>>> {
-    // generate random point in the field set
-    let assignments = generate_random_assignment(queriables);
-
-    // hash table with the hash of the expression as the key
-    // and the weak reference to the expression as the value
-    let mut seen_hashes: HashMap<u64, Weak<Expr<F, V, HashResult>>> = HashMap::new();
-
-    let mut result = Vec::new();
-
-    for expr in exprs {
-        let hashed_expr = Rc::new(expr.hash(&assignments));
-        let hash = hashed_expr.meta().hash;
-
-        // if the hash is already in the hash table,
-        // push the existing expression to the result
-        if let Some(existing_weak) = seen_hashes.get(&hash) {
-            if let Some(existing_expr) = existing_weak.upgrade() {
-                result.push(existing_expr);
-                continue;
-            }
-        }
-
-        // otherwise, insert the hash and the weak reference to the expression
-        seen_hashes.insert(hash, Rc::downgrade(&hashed_expr));
-        result.push(hashed_expr);
-    }
-
-    result
-}
-
-fn generate_random_assignment<F: Field + Hash, V: Debug + Clone + Eq + Hash>(
-    queriables: &Vec<V>,
-) -> VarAssignments<F, V> {
-    let mut rng = ChaCha20Rng::seed_from_u64(0);
+use super::{ConstrDecomp, Expr, HashResult, SignalFactory};
+use std::{fmt::Debug, hash::Hash};
 
-    let mut assignments = HashMap::new();
+/// This function replaces a common subexpression in an expression with a new signal.
+pub fn replace_expr<F: Field + Hash, V: Clone + Eq + Hash + Debug>(
+    expr: &Expr<F, V, HashResult>,
+    common_se: &Expr<F, V, HashResult>,
+) -> Expr<F, V, HashResult> {
+    let common_expr_hash = common_se.meta().hash;
 
-    for queriable in queriables {
-        let value = F::random(&mut rng);
-        assignments.insert(queriable.clone(), value);
+    if expr.meta().degree < common_se.meta().degree {
+        // If the current expression's degree is less than the common subexpression's degree,
+        // it can't contain the common subexpression, so we return it as is
+        expr.clone()
+    }
+    // If the expression is the same as the common subexpression return the signal
+    else if expr.meta().hash == common_expr_hash {
+        common_se.clone()
+    } else {
+        // Recursively apply the function to the subexpressions
+        expr.apply_subexpressions(|se| replace_expr(se, common_se))
     }
+}
 
-    assignments
+/// This function creates a new signal for a common subexpression.
+pub fn create_common_ses_signal<
```

Comment:
why do we need to return a ConstrDecomp? if we jsut return the signal is ok, because anyway outside you get what you put in ConstrDecomp and do other stuff with it

Reply:
can you check this comment pls regarding this? https://github.com/privacy-scaling-explorations/chiquito/pull/264#discussion_r1721563179